### ISSUE TITLE
Add fitness proportional selection

### ISSUE THREAD

**Issue by thomashopkins32** [AureumChaos/LEAP#143]
Implement fitness proportional selection as:

`select_prob(x,t) = fitness(x,t) / population_fitness(t)`

where population fitness is the sum of each `fitness(x,t)` in the population.

To avoid negative fitness we need to shift the each fitness by the minimum of the population:

`fitness(x,t) = optimality(x) + pop_min(t)`

Then randomly select individuals based on `select_prob(x,t)`.

Considering adding options such as:
- raising to a power (`fitness(x,t)^p`) where `p` is an integer. This can change the fitness differentials among the population which changes the selection pressure without changing the selection operator. Default will be `p=1`.
- smoothness by preserving `pop_min(t)` over previous `g` generations. For small populations, the `pop_min(t)` can vary a lot in each generation. This will allow the algorithm to keep track of the minimum observed fitness over the past `g` generations. Default will be `g=None` where minimum observed is calculated over all generations. Traditional fitness proportional selection would have `g=1`.

Also plan to add [stochastic universal sampling (SUS)](https://en.wikipedia.org/wiki/Stochastic_universal_sampling) but this can be a separate issue.

**Comment by thomashopkins32** [comment]
Currently unsure about allowing for addition of a constant such as `pop_min(t)` to all fitness values is a good idea.

This allows for handling the case with negative fitness values but creates weird behavior with all positive fitness values.

For example,
```
Individuals A and B
fitness(A) = 5
fitness(B) = 10
fitness_proportional_selection([A, B]) should produce [5/15, 10/15] or [0.333, 0.666] as probabilities.

Add Individual C
fitness(C) = -1000
If we try to add a constant of 1000 to each's fitness we get
fitness(A) = 1005
fitness(B) = 1010
fitness(C) = 0
This leaves C with 0% chance of being selected and 
makes selection of the other two almost equal.fitness_proportional_selection([A, B, C]) 
should produce [1005/2015, 1010/2015, 0/2015] or [0.499, 0.501, 0.0] as probabilities.
```

and with only positive values if we add the `pop_min(t)`

```
Individuals A and B
optimality(A) = 5
optimality(B) = 10
pop_min(t) = 5
fitness(A) = 5 + 5 = 10
fitness(B) = 10 + 5 = 15
fitness_proportional_selection([A, B]) should produce [10/25, 15/25] or [0.4, 0.6] as probabilities.
This is clearly different from the expected proportions [0.333, 0.666]
```

**Comment by thomashopkins32** [comment]
ECJ does not allow for negative fitness values, which would eliminate the need to add a constant such as `pop_min(t)` https://cs.gmu.edu/~eclab/projects/ecj/docs/

**Comment by SigmaX** [comment]
What about having an argument that allows users to set the "zero point" for purposes of fitness-prop comparisons?

Something like 

```python
ops.select_fitness_prop(zero_point=-1)
```

This would default to `zero_point=0`, and we can add error handling to throw a `ValueError` if any individual has `ind.fitness < zero_point`.

The behavior you originally defined could be achieved by using a special string argument:

```python
ops.select_fitness_prop(zero_point='pop-min')
```

**Comment by thomashopkins32** [comment]
This seems like the right thing to do since we want to be as flexible as possible.

So if I am understanding this correctly, this could be an `offset` from the zero point and if any value after adding the offset is below zero then we throw a `ValueError`.

Is `zero_point` or `offset` more intuitive for this case?

`offset = -zero_point` either way.

**Comment by SigmaX** [comment]
Yeah, sounds good.

Personally I find `zero_point` easier to think about mathematically.  But I like the name `offset` better, lol.  Up to you.

**Comment by thomashopkins32** [comment]
Issue resolved as of #144
